animation_nodes/nodes/sound/MIDI_Note_Evaluate.py

```diff
@@ -0,0 +1,61 @@
+import bpy
+from ... data_structures import DoubleList
+from ... base_types import AnimationNode, ListTypeSelectorSocket, VectorizedSocket
+
+def filterByNote(self, notesIn, noteNumber, time, attackTime, attackInterpolation, releaseTime, releaseInterpolation):
+    if self.useNoteNumberList:
+        notesInTime = [note for note in notesIn for n in noteNumber if
+            (time >= note.timeOn and time <= (note.timeOff + releaseTime) and note.noteNumber == n)]
+        notesTmp = list(set([note.noteNumber for note in notesInTime]))
+        if len(notesTmp) == 0:
+            numberTmp = [0.0 for n in noteNumber]
+            return DoubleList.fromValues(numberTmp)
+        numberTmp = []
+        for n in noteNumber:
+            noteTested = [notenum for notenum in notesTmp if notenum == n]
+            if len(noteTested) != 0:
+                notePlayed = [note for note in notesInTime if note.noteNumber == n]
+                notePlayed = notePlayed[0]
+                if time >= (notePlayed.timeOn + attackTime) and time <= notePlayed.timeOff:
+                    numberTmp.append(1.0)
+                elif time >= notePlayed.timeOn and time <= (notePlayed.timeOn + attackTime):
+                    value = (time - notePlayed.timeOn) / attackTime
+                    value = attackInterpolation(value)
+                    numberTmp.append(value)
+                elif time >= notePlayed.timeOff and time <= (notePlayed.timeOff + releaseTime):
+                    value = 1-((time - notePlayed.timeOff) / releaseTime)
+                    value = releaseInterpolation(value)
+                    numberTmp.append(value)
+            else:
+                numberTmp.append(0.0)
+        return DoubleList.fromValues(numberTmp)
+    else:
+        notesTmp = [note for note in notesIn if
```

Comment:
There should be a single function that does the evaluation. The list should be generated by multiple calls to this function. It is unclear to me why the code is structured like this.

Reply:
I'm not shure about what you said. Maybe you want I have two execute function, one when noteNumber is a single number and one execute function if noteNumber is a list ?

Reply:
Lets say the function F(N) evaluates the notes at some note number N. Then to evaluate a list of note numbers L, we should do `[F(N) for N in L]`. In your case, you seem to have two very different algorithms for computing a single and multiple note numbers.

Reply:
It has been a big and very interesting challenge for me. I hope the result will be good enought.